extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,371 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {evaluateBindExpr} from './bind-expr';
+import {getMode} from '../../../src/mode';
+import {isExperimentOn} from '../../../src/experiments';
+import {user} from '../../../src/log';
+import {vsyncFor} from '../../../src/vsync';
+
+const TAG = 'AMP-BIND';
+
+/**
+ * A single binding, e.g. <element [property]="expression"></element>.
+ * @typedef {{
+ *   property: !string,
+ *   expression: !string,
+ *   element: !Element
+ * }}
+ */
+let BindingDef;
+
+/**
+ * The state passed through vsync measure/mutate during a Bind digest cycle.
+ * @typedef {{
+ *   results: Array<BindExpressionResultDef>,
+ *   verifyOnly: boolean
+ * }}
+ */
+let BindVsyncStateDef;
+
+/**
+ * Possible types of a Bind expression evaluation.
+ * @typedef {(null|boolean|string|number|Array|Object)}
+ */
+let BindExpressionResultDef;
+
+/**
+ * Bind is the service that handles the Bind lifecycle, from identifying
+ * bindings in the document to scope mutations to reevaluating expressions
+ * during a digest.
+ */
+export class Bind {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @const {boolean} */
+    this.enabled_ = isExperimentOn(ampdoc.win, TAG);
+    user().assert(this.enabled_, `Experiment "${TAG}" is disabled.`);
+
+    /** @const {!../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampdoc = ampdoc;
+
+    /** {!Array<BindingDef>} */
+    this.bindings_ = [];
+
+    /**
+     * Expression results during the previous digest cycle, where the i'th
+     * element is the last result of `this.bindings_[i].expression`.
+     * @type {!Array<BindExpressionResultDef>}
+     */
+    this.previousResults_ = [];
+
+    /** @const {!Object} */
+    this.scope_ = Object.create(null);
+
+    /** @const {!../../../src/service/vsync-impl.Vsync} */
+    this.vsync_ = vsyncFor(ampdoc.win);
+
+    /** @const {!Function} */
+    this.boundMeasure_ = this.measure_.bind(this);
+
+    /** @const {!Function} */
+    this.boundMutate_ = this.mutate_.bind(this);
+
+    this.ampdoc.whenBodyAvailable().then(body => {
+      this.bindings_ = this.scanForBindings_(body);
```

Comment:
This alone is not enough and can't handle dynamic content (e.g. rendered via amp-mustache templates). Ideally we could use MutationObserver to run this on newly added DOM but aside from not being supported everywhere, I am not sure of its performance. Maybe a good compromise it to have the `template` service either run this on render of new templates or it triggers an event we can listen to here to scan again (I like the latter for looser coupling)